File: ios/Cookpit/Cookpit/MapDataController.swift
// //  MapDataController.swift
//  Cookpit
//
//  Created by Kittinun Vantasin on 7/15/16.
//  Copyright © 2016 Cookpit. All rights reserved.
//

import Foundation
import RxSwift

class MapDataController {
  
  let controller: CPMapController
  
  lazy var mapToken: String = { CPMapController.mapToken() }()

  fileprivate let _viewData = Variable<CPMapViewData?>(nil)
  
  lazy var viewData: Observable<CPMapViewData> = {
    self._viewData.asObservable()
        .filter { $0 != nil }
        .map { $0! as CPMapViewData }
        .observeOn(MainScheduler.instance)
  }()
  
  init() {
    controller = CPMapController.create()!
    controller.subscribe(self)
  }
  
  func unsubscribe() {
    controller.unsubscribe()
  }
  
  func request() {
    controller.request()
  }

}

extension MapDataController : CPMapControllerObserver {

  func onBeginUpdate() {

  }

  func onUpdate(_ data: CPMapViewData) {
    _viewData.value = data
  }

  func onEndUpdate() {
      
  }

}

File: android/Cookpit/app/src/main/kotlin/com/github/kittinunf/cookpit/map/MapDataController.kt
package com.github.kittinunf.cookpit.map

import com.github.kittinunf.cookpit.MapController
import com.github.kittinunf.cookpit.MapControllerObserver
import com.github.kittinunf.cookpit.MapViewData
import com.github.kittinunf.reactiveandroid.MutableProperty

class MapDataController : MapControllerObserver() {

    private val controller = MapController.create()

    private val _viewData = MutableProperty<MapViewData>()

    val viewData by lazy { _viewData.observable }

    companion object {
        fun mapToken() = MapController.mapToken()
    }

    init {
        controller.subscribe(this)
    }

    fun request() {
        controller.request()
    }

    fun unsubscribe() {
        controller.unsubscribe()
    }

    //region MapControllerObserver
    override fun onBeginUpdate() {
    }

    override fun onUpdate(viewData: MapViewData?) {
        _viewData.value = viewData
    }

    override fun onEndUpdate() {
    }
    //endregion
}